### PR TITLE
Fix mypy typing for `utilities.cloud_io.py`

### PR DESCRIPTION
**Pull request by stancld**
## What does this PR do?
Part of #7037. This PR disentangles large #7218. It fixes mypy typing in utilities.cloud_io.py.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/utilities/cloud_io.py [hunk 1]
```diff
@@ -14,15 +14,18 @@
...
 import torch
-from fsspec.implementations.local import LocalFileSystem
+from fsspec.implementations.local import AbstractFileSystem, LocalFileSystem
 from packaging.version import Version
 
 
-def load(path_or_url: Union[str, IO, Path], map_location=None):
+def load(
+    path_or_url: Union[str, IO, Path],
+    map_location: Optional[Union[str, torch.device, Dict[Union[str, torch.device], Union[str, torch.device]]]] = None,
```
Comment (awaelchli) [inline pytorch_lightning/utilities/cloud_io.py:27]: this can be a callable too in case of torch.load

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/cloud_io.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/utilities/cloud_io.py]: Done !

File: pytorch_lightning/utilities/cloud_io.py [hunk 2]
```diff
@@ -42,7 +47,7 @@ def get_filesystem(path: Union[str, Path]):
     return LocalFileSystem()
 
 
-def atomic_save(checkpoint, filepath: str):
+def atomic_save(checkpoint: Any, filepath: str) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/utilities/cloud_io.py:50]: is `checkpoint` Any or Dict[str, Any] ? 

similarly: is `filepath` `str` or `Union[str, Path]` ?

Comment (stancld) [reply pytorch_lightning/utilities/cloud_io.py:50]: @ananthsub According to the original doc, `filepath` for `fsspec.open` should be either `str` or `list` but `fsspec.Path` seems to me to work fine as well.
I agree, `checkpoint` should be rather `Dict[str, Any]` as you suggested as well.
